### PR TITLE
Firebase Services

### PR DESCRIPTION
**Pull request by rickbau5**
Scrape product pages for all Firebase related products. There is some overlap between what is scraped from the gcp script and from this new script. 

TODO
- [x] scrape Firebase services from product pages
- [ ] include services in `data/gcp.json`

### CODE REVIEW

File: data/gcp-firebase.json [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+      }
+    ],
+    "tags": [
+      "gcp/platform",
+      "gcp/firebase",
+      "gcp/firebase/dynamic-links"
+    ]
+  },
+  {
+    "id": "gcp/firebase-firebase-extensions",
```
Comment (rickbau5) [inline data/gcp-firebase.json:173]: this happens when the product `name` has `Firebase` in it already. Can be removed from the id.

**Comment by rickbau5** [comment]
The script is not including these in the `data/gcp.json` yet. Wanted to get an initial review of the approach here and make some calls about the following cases that need to be handled.

* Cloud Storage & Cloud Functions exist as a product offering within Firebase as well as outside as part of the normal GCP offerings
* Crashlytics, Test Lab, and Firestore Realtime Database appear both on the overall GCP product listing as well as within the Firebase offerings. For some reason these are called out specifically in the GCP product listing whereas the others are not.

<details>
<summary>
Cloud Storage for Firebase
</summary>

From GCP Firebase services
```json
{
  "id": "gcp/firebase-cloud-storage",
  "name": "Cloud Storage",
  "summary": "Store and serve user-generated content with ease as your app grows from prototype to production-ready.",
  "url": "https://firebase.google.com/products/storage",
  "categories": [
    {
      "id": "firebase",
      "name": "Firebase"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/category/firebase",
    "gcp/service/firebase-cloud-storage"
  ]
}
```
From GCP services
```json
{
  "id": "gcp/cloud-storage-for-firebase",
  "name": "Cloud Storage for Firebase",
  "summary": "Object storage for storing and serving user-generated content.",
  "url": "https://firebase.google.com/products/storage/",
  "categories": [
    {
      "id": "storage",
      "name": "storage"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/service/cloud-storage-for-firebase",
    "gcp/category/storage"
  ]
}
```

</details>

<details>
<summary>
Firebase Test Lab
</summary>

From GCP Firebase Services
```json
{
  "id": "gcp/firebase-test-lab",
  "name": "Test Lab",
  "summary": "Spot errors before release by testing and validating your app on physical and virtual devices that simulate actual environments.",
  "url": "https://firebase.google.com/products/test-lab",
  "categories": [
    {
      "id": "firebase",
      "name": "Firebase"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/category/firebase",
    "gcp/service/firebase-test-lab"
  ]
}
```
From GCP services
```json
{
  "id": "gcp/firebase-test-lab",
  "name": "Firebase Test Lab",
  "summary": "On-demand testing infrastructure for Android apps.",
  "url": "https://firebase.google.com/docs/test-lab/",
  "categories": [
    {
      "id": "developer-tools",
      "name": "developer tools"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/service/firebase-test-lab",
    "gcp/category/developer-tools"
  ]
}
```
</details>

<details>
<summary>
Firebase Realtime Database
</summary>

From GCP Firebase services
```json
{
  "id": "gcp/firebase-realtime-database",
  "name": "Realtime Database",
  "summary": "Build serverless apps by storing and syncing JSON data between your users in near-realtime, on or offline, with strong user-based security.",
  "url": "https://firebase.google.com/products/realtime-database",
  "categories": [
    {
      "id": "firebase",
      "name": "Firebase"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/category/firebase",
    "gcp/service/firebase-realtime-database"
  ]
}
```
From GCP services
```json
{
  "id": "gcp/firebase-realtime-database",
  "name": "Firebase Realtime Database",
  "summary": "NoSQL database for storing and syncing data in real time.",
  "url": "https://firebase.google.com/products/realtime-database/",
  "categories": [
    {
      "id": "databases",
      "name": "databases"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/service/firebase-realtime-database",
    "gcp/category/databases"
  ]
}
```
</details>

<details>
<summary>
Firebase Crashlytics
</summary>

From GCP Firebase services
```json
{
  "id": "gcp/firebase-crashlytics",
  "name": "Crashlytics",
  "summary": "Track, prioritize and fix stability issues that erode app quality, in realtime.",
  "url": "https://firebase.google.com/products/crashlytics",
  "categories": [
    {
      "id": "firebase",
      "name": "Firebase"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/category/firebase",
    "gcp/service/firebase-crashlytics"
  ]
}
```
From GCP services
```json
{
  "id": "gcp/firebase-crashlytics",
  "name": "Firebase Crashlytics",
  "summary": "Prioritize and fix stability issues faster.",
  "url": "https://firebase.google.com/products/crashlytics/",
  "categories": [
    {
      "id": "developer-tools",
      "name": "developer tools"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/service/firebase-crashlytics",
    "gcp/category/developer-tools"
  ]
}
```
</details>

<details>
<summary>
Cloud Functions
</summary>

From GCP Firebase services
```json
{
  "id": "gcp/firebase-cloud-functions",
  "name": "Cloud Functions",
  "summary": "Write and run app logic server-side without needing to set up your own server.",
  "url": "https://firebase.google.com/products/functions",
  "categories": [
    {
      "id": "firebase",
      "name": "Firebase"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/category/firebase",
    "gcp/service/firebase-cloud-functions"
  ]
}
```

From GCP services
```json
{
  "id": "gcp/cloud-functions",
  "name": "Cloud Functions",
  "summary": "Event-driven compute platform for cloud services and apps.",
  "url": "https://cloud.google.com/functions",
  "categories": [
    {
      "id": "featured-products",
      "name": "featured products"
    },
    {
      "id": "serverless-computing",
      "name": "serverless computing"
    }
  ],
  "tags": [
    "gcp/platform",
    "gcp/service/cloud-functions",
    "gcp/category/featured-products",
    "gcp/category/serverless-computing"
  ]
}
```

</details>

**Comment by bkw** [comment]
Great stuff, @rickbau5, thanks for your contribution!
We should probably filter out the conflicting entries from gcp , except for cloud functions, I think that's an acceptable duplication. We also should not forget to tell CREs about this new firebase category once this is merged.

@alexei-led , do you agree that the various firebase related services warrant a firebase category?

**Comment by rickbau5** [comment]
Hey thanks @bkw - the request to add the firebase services did come internally. I will connect with you and share the motivation. I wanted to get this put up so we could see what services we could scrape and make a determination of what we want to service.

Thanks for taking the time to look!

**Comment by arri-cc** [comment]
There's a couple of nuances here:

- Firebase Cloud Functions and Firebase Cloud Storage are both backed by the cloud services of the same names, but are exposed through a different API for convenience (user authentication, web hosting, etc).  Also, cloud functions I'd limited to JavaScript as the only language that's supported.
- Crashlytics, Firestore and Test Lab likely fall under both GCP and Firebase because they're all officially considered cloud products, but we should confirm this point.

These tags will help us know how to support customers and truly measure how often we're actually supporting the Firebase product suite compared to others.

**Comment by arri-cc** [comment]
ran it through chatgpt

`gcp-firebase.py`
```python
import requests
import json
from bs4 import BeautifulSoup

pages = [
    "https://firebase.google.com/products-build",
    "https://firebase.google.com/products-release",
    "https://firebase.google.com/products-engage"
]

output_data = []

for page in pages:
    response = requests.get(page)
    soup = BeautifulSoup(response.text, 'html.parser')
    divs = soup.select('div.product-grid__content')

    for div in divs:
        item = {}
        a_tag = div.select_one("a")
        p_tag = div.select_one("p")

        href = a_tag.get("href", "")
        item_id = a_tag.get("id", "").replace(" ", "-").replace("firebase", "").lstrip("-")

        item["id"] = f"gcp/firebase-{item_id}"
        item["name"] = a_tag.text.strip()
        item["summary"] = p_tag.text.strip()
        item["url"] = href if href.startswith("https://") else f"https://firebase.google.com{href}"
        item["categories"] = [{"id": "firebase", "name": "Firebase"}]
        item["tags"] = [
            "gcp/platform",
            "gcp/category/firebase",
            f"gcp/service/firebase-{item_id}"
        ]

        output_data.append(item)

# Remove duplicates and sort by id
output_data = sorted({d['id']: d for d in output_data}.values(), key=lambda x: x['id'])

with open("data/gcp-firebase.json", "w") as f:
    json.dump(output_data, f, indent=4)

print("done")
```

and then `tags.py`

```python
import json

# List of input JSON files
input_files = [
    "data/aws.json",
    "data/gcp.json",
    "data/cmp.json",
    "data/gsuite.json",
    "data/ms365.json",
    "data/azure.json",
    "data/finance.json",
    "data/credits.json",
    "data/gcp-firebase.json"
]

all_tags = []

# Loop through each file to collect all tags
for file in input_files:
    with open(file, 'r') as f:
        data = json.load(f)
        for item in data:
            if 'tags' in item:
                all_tags.extend(item['tags'])

# Remove duplicates and sort
all_tags = sorted(set(all_tags))

# Save to output file
with open("data/tags.json", 'w') as f:
    json.dump(all_tags, f, indent=4)

print("done")
```

**Pull request closed without merging**